Hazel/src/Hazel/Debug/Instrumentor.h

```diff
@@ -124,10 +124,17 @@ namespace Hazel {
 
 #define HZ_PROFILE 1
 #if HZ_PROFILE
+	#if defined(__FUNCSIG__)
+		#define HZ_FUNC_SIG __FUNCSIG__
+	#elif defined(__PRETTY_FUNCTION__)
+		#define HZ_FUNC_SIG __PRETTY_FUNCTION__
+	#else
+		#define HZ_FUNC_SIG __FUNCTION__
+	#endif
```

Comment:
Should we first check to see if `__func__` is available and use it before using `__FUNCTION__`? See [this stackoverflow answer/post](https://stackoverflow.com/a/4384825) for the reasoning.

Reply:
According to [the language references](https://docs.microsoft.com/en-us/cpp/cpp/func?view=vs-2019), cpp says __func__ is mandated since C++11 (been mandated in C99 as well).

With that note, do we need any macro definitions then? Should we just use __func__ instead as this should be cross platform, as we are compiling with C++17 anyway?

Reply:
I believe the string returned by `__func__` is less detailed than `__FUNCSIG__` and `__PRETTY_FUNCTION__` ([on gcc at least](https://godbolt.org/z/edtdmm)) so we should use those when available. We can remove the case for `__FUNCTION__` though, since (as you said) we are compiling with C++17.

Reply:
I guess now we have to do some research about how to find out what compiler we're using and such, what functions are available...

Reply:
After alot of research I came up with this solution, mainly based on [Boost's assert function](https://www.boost.org/doc/libs/1_66_0/libs/assert/doc/html/assert.html#current_function_macro_boost_current_function_hpp):

```cpp
#if defined(__GNUC__) || (defined(__MWERKS__) && (__MWERKS__ >= 0x3000)) || (defined(__ICC) && (__ICC >= 600)) || defined(__ghs__)
	# define HZ_FUNC_SIG __PRETTY_FUNCTION__
#elif defined(__DMC__) && (__DMC__ >= 0x810)
	# define HZ_FUNC_SIG __PRETTY_FUNCTION__
#elif defined(__FUNCSIG__)
	# define HZ_FUNC_SIG __FUNCSIG__
#elif (defined(__INTEL_COMPILER) && (__INTEL_COMPILER >= 600)) || (defined(__IBMCPP__) && (__IBMCPP__ >= 500))
	# define HZ_FUNC_SIG __FUNCTION__
#elif defined(__BORLANDC__) && (__BORLANDC__ >= 0x550)
	# define HZ_FUNC_SIG __FUNC__
#elif defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901)
	# define HZ_FUNC_SIG __func__
#elif defined(__cplusplus) && (__cplusplus >= 201103)
	# define HZ_FUNC_SIG __func__
#else
	# define HZ_FUNC_SIG "HZ_FUNC_SIG unknown!"
#endif
```

What do you think @Puddlestomper ?

Reply:
Looks fine to me. I don't know all (or even most) of these macros, but it seems to work for gcc and clang on Godbolt and I confirmed it works for Visual Studio and gcc/g++ in test projects.

If there are any issues with this then we can always change it. For now it will at least work for the main compilers so I think it's good. Thanks for doing the research to figure this out.

Reply:
Resolved in https://github.com/TheCherno/Hazel/pull/174/commits/2b06109be382a435724c782d9f69ef8e228ec672